### PR TITLE
kv: add replicas.uninitialized metric

### LINKED ISSUE #73368: kv: add replicas.uninitialized metric
In a recent customer cluster, we noticed that it would be nice to have a time-series metric that tracked the number of uninitialized replicas on a node.

We should add one called `replicas.uninitialized`, which will behave similarly to `replicas.quiescent` in that it will be updated in `Store.updateReplicationGauges`, but will track the size of the `s.mu.uninitReplicas` map. This map doesn't quite give us the number of uninitialized replicas because, after a process restart, uninitialized replicas can lie dormant in persistent state without any corresponding in-memory representation. However, it's a close enough approximation and accounts for the more expensive form of uninitialized replicas. We'll just want to mention this caveat in the metric description.

### PR DESCRIPTION
**Pull request by kvoli**
This change adds `replicas.unitialized` metric, tracking the number of non-init replicas on a store. 

Note that this metric doesn't give the exact number of uninitialized replicas because, after a process restart, uninitialized replicas can remain on disk without an in memory representation.

resolves  #73368

### CODE REVIEW

File: pkg/kv/kvserver/store.go [hunk 1]
```diff
@@ -3011,6 +3012,9 @@ func (s *Store) updateReplicationGauges(ctx context.Context) error {
 		if metrics.Quiescent {
 			quiescentCount++
 		}
+
+		uninitializedCount = int64(len(s.mu.uninitReplicas))
```
Comment (irfansharif) [inline pkg/kv/kvserver/store.go:3016]: Weird, github reviews don't show where this map is being updated, but I could see it locally. In any case, some notes from when we looked at this offline:

- The per-replica visitor above (`newStoreReplicaVisitor`) also iterates over uninitialized replicas. Instead of tracking+maintaining the set of uninitialized replicas in a separate map (probably more error prone), using the existing replica iteration + checking if the current replica is initialized is I think sufficient.
- We'd want a test. We looked at another test to steal from where we could block the reception of a snapshot on a second node, ensure that our uninitialized count is as we'd expect, unblock the replica initialization, and ensure it's back to 0.

File: pkg/kv/kvserver/metrics.go [hunk 2]
```diff
@@ -68,6 +68,12 @@ var (
 		Measurement: "Replicas",
 		Unit:        metric.Unit_COUNT,
 	}
+	metaUninitializedCount = metric.Metadata{
+		Name:        "replicas.uninitialized",
+		Help:        "Number of uninitialized replicas, this does not include uninitialized replicas that can lie dormant in a persistent state.",
```
Comment (irfansharif) [inline pkg/kv/kvserver/metrics.go:73]: > does not include uninitialized replicas that can lie dormant in a persistent state

Interesting -- Mind explaining how? Are uninitialized replicas ever persisted?

**Review by irfansharif — COMMENTED** [review]
Sweet turnaround time! And congrats on your first PR! In addition to the comments below, let's squash these changes into a single commit. We have [this wiki-page](https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages) on good commit hygiene that you might find interesting. For here I'd just suggest rewording the title to read something similar to what you have for the PR: "kvserver: add metric for uninitialized replicas".

**Comment by irfansharif** [comment]
Note on labels: for most of them (mod `do-not-merge`, `X-noremind`, etc) we really only use them for filed issues.

File: pkg/kv/kvserver/metrics.go (hunk 2 above)
Comment (kvoli) [reply pkg/kv/kvserver/metrics.go:73]: This is from: https://github.com/cockroachdb/cockroach/issues/73368. @nvanbenschoten mentions 

> uninitialized replicas can lie dormant in persistent state without any corresponding in-memory representation

File: pkg/kv/kvserver/client_replica_test.go [hunk 3]
```diff
@@ -3901,6 +3901,123 @@ func TestTenantID(t *testing.T) {
...
 
+//  TestUninitializedMetric
+//  This test examines the following behaviors:
+//  (1): When a replica is in an uninitialized state, the store containing that replica 
+//       should reflect this in store metric: UninitializedCount
+func TestUninitializedMetric(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	stickyEngineRegistry := server.NewStickyInMemEnginesRegistry()
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:3912]: I think we can get rid of all usages of `stickyEngineRegistry`, `raftConfig`, and `stickySpecTestServerArgs`. We're not restarting an existing test server and don't need server state to be persisted across the restarts, which is what all this machinery sets out to do. Trying it out locally, all we really need is the following:

```
	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
		ReplicationMode: base.ReplicationManual,
		ServerArgs:      base.TestServerArgs{Insecure: true},
	})
```

And below:

```
	tc.AddAndStartServer(t, base.TestServerArgs{
		Insecure: true,
```

File: pkg/kv/kvserver/client_replica_test.go [hunk 4]
```diff
@@ -3901,6 +3901,123 @@ func TestTenantID(t *testing.T) {
...
+			},
+		},
+	}
+	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
+		ReplicationMode: base.ReplicationManual,
+		ServerArgs:      stickySpecTestServerArgs,
+	})
+	defer tc.Stopper().Stop(ctx)
+
+	tenant2 := roachpb.MakeTenantID(2)
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:3945]: While we technically _can_ split out a separate range using a tenant prefix, we can do something more direct:

```go
	scratchKey := tc.ScratchRange(t)
	_, repl := getFirstStoreReplica(t, tc.Server(0), scratchKey)
```

File: pkg/kv/kvserver/client_replica_test.go [hunk 5]
```diff
@@ -3901,6 +3901,123 @@ func TestTenantID(t *testing.T) {
...
+		sawSnapshot := make(chan struct{}, 1)
+		blockSnapshot := make(chan struct{})
+		tc.AddAndStartServer(t, base.TestServerArgs{
+			RaftConfig: raftConfig,
+			Insecure:   true,
+			Knobs: base.TestingKnobs{
+				Store: &kvserver.StoreTestingKnobs{
+					BeforeSnapshotSSTIngestion: func(
+						snapshot kvserver.IncomingSnapshot,
+						request_type kvserver.SnapshotRequest_Type,
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:3960]: [nit] Given these are unused, `s/request_type/_` and `s/strings/_`.

File: pkg/kv/kvserver/client_replica_test.go [hunk 6]
```diff
@@ -3901,6 +3901,123 @@ func TestTenantID(t *testing.T) {
...
+		ReplicationMode: base.ReplicationManual,
+		ServerArgs:      stickySpecTestServerArgs,
+	})
+	defer tc.Stopper().Stop(ctx)
+
+	tenant2 := roachpb.MakeTenantID(2)
+	tenant2Prefix := keys.MakeTenantPrefix(tenant2)
+	// Ensure that a range with a tenant prefix has the proper tenant ID.
+	tc.SplitRangeOrFatal(t, tenant2Prefix)
+	t.Run("(1) uninitialized replic is reflected in UninitializedMetric", func(t *testing.T) {
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:3949]: Since there's only one sub-test, let's get rid of this `t.Run` outdent and make it part of the main test itself. We typically use sub-tests when testing minor alterations of a single test run (perhaps with shared setup), which isn't quite the case here.

File: pkg/kv/kvserver/client_replica_test.go [hunk 7]
```diff
@@ -3901,6 +3901,123 @@ func TestTenantID(t *testing.T) {
...
+				return nil
+			case err := <-addReplicaErr:
+				go addReplica()
+				return err
+			}
+		}); err != nil {
+			t.Fatal(err)
+		}
+
+		const forceMetricTick = iota
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:3997]: s/iota/0

File: pkg/kv/kvserver/client_replica_test.go [hunk 8]
```diff
@@ -3901,6 +3901,123 @@ func TestTenantID(t *testing.T) {
...
+		uninitStore := tc.GetFirstStoreFromServer(t, 1)
+
+		// Force the store to compute the replica metrics
+		err := uninitStore.ComputeMetrics(ctx, forceMetricTick)
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:4001]: [nit] Inline things here and below: 
```go
require.NoError(t, uninitStore.ComputeMetrics(ctx, 0))
// ...
require.Equal(t, int64(1), uninitStore.Metrics().UninitializedCount.Value())
// ...
require.NoError(t, uninitStore.ComputeMetrics(ctx, 0))
// ...
require.Equal(t, int64(0), uninitStore.Metrics().UninitializedCount.Value()) // now initialized
```

File: pkg/ui/workspaces/cluster-ui/src/store/nodes/nodes.fixtures.ts [hunk 9]
```diff
@@ -888,6 +888,7 @@ export const getNodeStatus = () => {
           "replicas.leaders_not_leaseholders": 0,
           "replicas.leaseholders": 77,
           "replicas.quiescent": 77,
+          "replicas.uninitialized": 0,
```
Comment (irfansharif) [inline pkg/ui/workspaces/cluster-ui/src/store/nodes/nodes.fixtures.ts:891]: Is this diff still relevant given we're not adding it to the cluster UI? (anything in addition to the advanced debug page i.e.)

**Review by irfansharif — COMMENTED** [review]
Nice test! I only have a few minor suggestions to trim it down and we should be all set. The CI failures also suggest an update to TestChartCatalogMetrics.

**Review by kvoli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73975)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/kv/kvserver/store.go, line 3016 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-Mr7umb6PL1Hj2HVN2lI-r4-3016:-MrXXCFC9YQQ9I__1kNH:b-3heoja) ([raw file](https://github.com/cockroachdb/cockroach/blob/413b9ff21c584fc52f5655877f5cb1255ea2f4c3/pkg/kv/kvserver/store.go#L3016)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Weird, github reviews don't show where this map is being updated, but I could see it locally. In any case, some notes from when we looked at this offline:

- The per-replica visitor above (`newStoreReplicaVisitor`) also iterates over uninitialized replicas. Instead of tracking+maintaining the set of uninitialized replicas in a separate map (probably more error prone), using the existing replica iteration + checking if the current replica is initialized is I think sufficient.
- We'd want a test. We looked at another test to steal from where we could block the reception of a snapshot on a second node, ensure that our uninitialized count is as we'd expect, unblock the replica initialization, and ensure it's back to 0.
</blockquote></details>

Done. We resolved to use the original method and reuse the test.
___
*[pkg/kv/kvserver/client_replica_test.go, line 3912 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r5-3912:-MrXeD2_5kIn90cCjG_9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78acdf22816c24c98f7de21a85f14ca45f3fea25/pkg/kv/kvserver/client_replica_test.go#L3912)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I think we can get rid of all usages of `stickyEngineRegistry`, `raftConfig`, and `stickySpecTestServerArgs`. We're not restarting an existing test server and don't need server state to be persisted across the restarts, which is what all this machinery sets out to do. Trying it out locally, all we really need is the following:

```
	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
		ReplicationMode: base.ReplicationManual,
		ServerArgs:      base.TestServerArgs{Insecure: true},
	})
```

And below:

```
	tc.AddAndStartServer(t, base.TestServerArgs{
		Insecure: true,
```
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_replica_test.go, line 3945 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r5-3945:-MrXeEFyAK3VRNCB6pdC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78acdf22816c24c98f7de21a85f14ca45f3fea25/pkg/kv/kvserver/client_replica_test.go#L3945)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

While we technically _can_ split out a separate range using a tenant prefix, we can do something more direct:

```go
	scratchKey := tc.ScratchRange(t)
	_, repl := getFirstStoreReplica(t, tc.Server(0), scratchKey)
```
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_replica_test.go, line 3949 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r5-3949:-MrXeFNt3ZeZPN8UJEjj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78acdf22816c24c98f7de21a85f14ca45f3fea25/pkg/kv/kvserver/client_replica_test.go#L3949)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Since there's only one sub-test, let's get rid of this `t.Run` outdent and make it part of the main test itself. We typically use sub-tests when testing minor alterations of a single test run (perhaps with shared setup), which isn't quite the case here.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_replica_test.go, line 3960 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r5-3960:-MrXeFuLB_B7mFc9RPUj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78acdf22816c24c98f7de21a85f14ca45f3fea25/pkg/kv/kvserver/client_replica_test.go#L3960)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Given these are unused, `s/request_type/_` and `s/strings/_`.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_replica_test.go, line 3997 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r5-3997:-MrXeGN64VF0MxZhNuP2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78acdf22816c24c98f7de21a85f14ca45f3fea25/pkg/kv/kvserver/client_replica_test.go#L3997)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/iota/0
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_replica_test.go, line 4001 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r5-4001:-MrXeGroBtImYHIjyc8M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78acdf22816c24c98f7de21a85f14ca45f3fea25/pkg/kv/kvserver/client_replica_test.go#L4001)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Inline things here and below: 
```go
require.NoError(t, uninitStore.ComputeMetrics(ctx, 0))
// ...
require.Equal(t, int64(1), uninitStore.Metrics().UninitializedCount.Value())
// ...
require.NoError(t, uninitStore.ComputeMetrics(ctx, 0))
// ...
require.Equal(t, int64(0), uninitStore.Metrics().UninitializedCount.Value()) // now initialized
```
</blockquote></details>

Done.
___
*[pkg/ui/workspaces/cluster-ui/src/store/nodes/nodes.fixtures.ts, line 891 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-Mr7umb6PL1Hj2HVN2lL-r5-891:-MrXeHcKEm0naouHqE_F:b-xgt6nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/78acdf22816c24c98f7de21a85f14ca45f3fea25/pkg/ui/workspaces/cluster-ui/src/store/nodes/nodes.fixtures.ts#L891)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Is this diff still relevant given we're not adding it to the cluster UI? (anything in addition to the advanced debug page i.e.)
</blockquote></details>

You are right, I thought it might be necessary to declare a default value. I tested without this line and it still shows for advanced debugging. Done.

**Review by kvoli — COMMENTED** [review]
@irfansharif thanks for the comments. I've adopted your suggestions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-:-MrXeTvM7aKSUu-PQPNZ:b-aziqoz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

File: pkg/kv/kvserver/client_replica_test.go [hunk 10]
```diff
@@ -3901,6 +3901,89 @@ func TestTenantID(t *testing.T) {
 
 }
 
+//  TestUninitializedMetric
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:3904]:
```
// TestUninitializedMetric ensures that uninitialized replicas are reflected in the UninitializedCount store metric.
```

File: pkg/kv/kvserver/client_replica_test.go [hunk 11]
```diff
@@ -3901,6 +3901,89 @@ func TestTenantID(t *testing.T) {
...
+	// Blocked snapshot on the second server (1) should realize 1 uninitialized replica.
+	require.Equal(t, int64(1), targetStore.Metrics().UninitializedCount.Value())
+
+	// Unblock snapshot on the second server (1), this should initialize replica.
+	// Again force the store to compute the metrics, increment tick counter
+	close(blockSnapshot)
+	forceMetricTick++
+
+	require.NoError(t, <-addReplicaErr)
+	require.NoError(t, targetStore.ComputeMetrics(ctx, forceMetricTick))
```
Comment (irfansharif) [inline pkg/kv/kvserver/client_replica_test.go:3981]: Btw, instead of maintaining the forceMetricCounter, just passing 0 directly in ComputeMetrics also works (looking at other usages that's what we seem to do).

**Review by irfansharif — APPROVED** [review]
Sweet, LGTM! For this kind of test I'd typically stress it locally to shake out any flakes, but because it's a new test our CI should already do that for you. For posterity:

```
$ dev test pkg/kv/kvserver -f=TestUninitializedMetric --stress
INFO: Found 1 test target
16 runs so far, 0 failures, over 5s
59 runs so far, 0 failures, over 10s
95 runs so far, 0 failures, over 15s
125 runs so far, 0 failures, over 20s
...
```

Had to first rebase on master to pick up #73982. Could also use `make stress PKG=./pkg/kv/kvserver TESTS=TestUninitializedMetric`.

**Review by nvb — COMMENTED** [review]
> Reviewed 3 of 5 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73975)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @kvoli)

___
*[-- commits, line 5 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrXu9Iy4CgQG9TkBduo):*
You'll want to link this to #73368 here and the in PR description. You have "Addresses" in the PR description, but unfortunately, that is not one of the supported keywords. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword.
___
*[pkg/kv/kvserver/metrics.go, line 73 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-Mr7umb6PL1Hj2HVN2lH-r4-73:-MrXuVVh-qzlu6bVrXnf:b-tgdu7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/413b9ff21c584fc52f5655877f5cb1255ea2f4c3/pkg/kv/kvserver/metrics.go#L73)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This is from: https://github.com/cockroachdb/cockroach/issues/73368. @nvanbenschoten mentions 

> uninitialized replicas can lie dormant in persistent state without any corresponding in-memory representation
</blockquote></details>

Right, there's a small amount of state (typically just the `HardState`) that can be persisted by an uninitialized replica. We don't look for this on process startup and don't instantiate a `Replica` object for these unless that replica hears a new message. Undercounting here is mostly fine, as this persisted state is tiny and we're mostly interested in the cost of these in-memory `Replica` objects.

There's some discussion about this in [here](https://github.com/cockroachdb/cockroach/blob/83078148e390fbea1c29eef064b9d1f95edc8d9a/pkg/kv/kvserver/store.go#L542) and [here](https://github.com/cockroachdb/cockroach/blob/83078148e390fbea1c29eef064b9d1f95edc8d9a/pkg/storage/replicas_storage.go).
___
*[pkg/kv/kvserver/metrics.go, line 1249 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrXvejt1bEwTCZONLkX:-MrXvejt1bEwTCZONLkY:baxv58c) ([raw file](https://github.com/cockroachdb/cockroach/blob/24bb005e36ea34adf386632aedb08595d711b4b4/pkg/kv/kvserver/metrics.go#L1249)):*
> ```Go
> 	RaftLeaderNotLeaseHolderCount *metric.Gauge
> 	LeaseHolderCount              *metric.Gauge
> 	UninitializedCount            *metric.Gauge
> ```

nit: keep the ordering consistent. So move this below `QuiescentCount`.
___
*[pkg/kv/kvserver/store.go, line 3016 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-Mr7umb6PL1Hj2HVN2lI-r4-3016:-MrXwHaN6rRs-xt2hi1A:b-m37siq) ([raw file](https://github.com/cockroachdb/cockroach/blob/413b9ff21c584fc52f5655877f5cb1255ea2f4c3/pkg/kv/kvserver/store.go#L3016)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Done. We resolved to use the original method and reuse the test.
</blockquote></details>

An alternative to this approach is to add a new `WithUninitialized()` method (similar to `InOrder`) to `storeReplicaVisitor` which avoids filtering out uninitialized replicas in `storeReplicaVisitor.Visit`. We could then add a small branch at the top of this visitor function that checks if the current replica is uninitialized (`(*Replica).IsInitialized`). If it is, increment `uninitializedCount` and `return true`.

But what we have here works too.
___
*[pkg/kv/kvserver/store.go, line 3019 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrXvooq7wjmkA-AsTgv:-MrXvooq7wjmkA-AsTgw:b-v3hadc) ([raw file](https://github.com/cockroachdb/cockroach/blob/24bb005e36ea34adf386632aedb08595d711b4b4/pkg/kv/kvserver/store.go#L3019)):*
> ```Go
> 			quiescentCount++
> 		}
> ```

nit: now that you've addressed Irfan's comment, there's a stray line here.

**Review by kvoli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73975)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @kvoli, and @nvanbenschoten)

___
*[-- commits, line 5 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrXu9Iy4CgQG9TkBduo):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You'll want to link this to #73368 here and the in PR description. You have "Addresses" in the PR description, but unfortunately, that is not one of the supported keywords. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword.
</blockquote></details>

Updated with a supported keyword.
___
*[pkg/kv/kvserver/metrics.go, line 73 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-Mr7umb6PL1Hj2HVN2lH-r4-73:-MrY4d-46VoRA3hucPxQ:begiopn) ([raw file](https://github.com/cockroachdb/cockroach/blob/413b9ff21c584fc52f5655877f5cb1255ea2f4c3/pkg/kv/kvserver/metrics.go#L73)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Right, there's a small amount of state (typically just the `HardState`) that can be persisted by an uninitialized replica. We don't look for this on process startup and don't instantiate a `Replica` object for these unless that replica hears a new message. Undercounting here is mostly fine, as this persisted state is tiny and we're mostly interested in the cost of these in-memory `Replica` objects.

There's some discussion about this in [here](https://github.com/cockroachdb/cockroach/blob/83078148e390fbea1c29eef064b9d1f95edc8d9a/pkg/kv/kvserver/store.go#L542) and [here](https://github.com/cockroachdb/cockroach/blob/83078148e390fbea1c29eef064b9d1f95edc8d9a/pkg/storage/replicas_storage.go).
</blockquote></details>

Noted, thanks for the context.
___
*[pkg/kv/kvserver/metrics.go, line 1249 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrXvejt1bEwTCZONLkX:-MrY45j1BYYb6RZ6VXIo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/24bb005e36ea34adf386632aedb08595d711b4b4/pkg/kv/kvserver/metrics.go#L1249)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: keep the ordering consistent. So move this below `QuiescentCount`.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store.go, line 3016 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-Mr7umb6PL1Hj2HVN2lI-r4-3016:-MrY4Oz2B2G4Cd1N0rbl:b8mshi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/413b9ff21c584fc52f5655877f5cb1255ea2f4c3/pkg/kv/kvserver/store.go#L3016)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

An alternative to this approach is to add a new `WithUninitialized()` method (similar to `InOrder`) to `storeReplicaVisitor` which avoids filtering out uninitialized replicas in `storeReplicaVisitor.Visit`. We could then add a small branch at the top of this visitor function that checks if the current replica is uninitialized (`(*Replica).IsInitialized`). If it is, increment `uninitializedCount` and `return true`.

But what we have here works too.
</blockquote></details>

So it is being filtered out here. I wasn't certain. I'll keep it as is atm, since I don't know what else it might be necessary for yet.
___
*[pkg/kv/kvserver/client_replica_test.go, line 3904 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r6-3904:-MrY4IB8-cBNT6AHOBJN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e72f50fb3ced6be46e92179a59f1ebee6292e28c/pkg/kv/kvserver/client_replica_test.go#L3904)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

```
// TestUninitializedMetric ensures that uninitialized replicas are reflected in the UninitializedCount store metric.
```
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_replica_test.go, line 3981 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73975#-MrTV053b82ycFhJonzx-r6-3981:-MrY4KBWDeZd1pBB-A5z:b7ci7rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e72f50fb3ced6be46e92179a59f1ebee6292e28c/pkg/kv/kvserver/client_replica_test.go#L3981)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Btw, instead of maintaining the forceMetricCounter, just passing 0 directly in ComputeMetrics also works (looking at other usages that's what we seem to do).
</blockquote></details>

Adopted this change.
